<a href="https://colab.research.google.com/github/oliveralbrecht94/ATDIT2/blob/master/Task%20B.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Import of needed classes**

In [0]:
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.manifold import TSNE
from scipy import stats
import scipy.spatial.distance


# Load MNIST DATASET

In [0]:
# load MNIST in mnistdataset
mnistdataset = tf.keras.datasets.mnist.load_data()[0]
# load first dimension (digits/images) in digitsand second dimension (labels) in labels
digits, labels = mnistdataset [0], mnistdataset [1]

In [0]:
# load MNIST in testdataset
testdataset = tf.keras.datasets.mnist.load_data()[1]
# load first dimension (digits) in test_digits and second dimension (labels) in test_labels
test_digits, test_labels = testdataset[0], testdataset[1]

# Preparing Dataset

In [0]:
# save amount of digits (from digits) in n_digits
n_digits = digits.shape[0] 
# save amount of digitis (from test_digits) in n_test
n_test = test_digits.shape[0] 

In [0]:
# calculation of full dimension of each image 28*28
amount_pixels = digits.shape[1] * digits.shape[2]

In [0]:
# Transform to 1D vector from each matrix with the pixels of an image in order to calculate distances.
digits = digits.reshape([n_digits, amount_pixels])
test_digits = test_digits.reshape([n_test, amount_pixels])

In [51]:
# see the result
digits.shape

(60000, 784)

In [0]:
# setting an index of 50.0000 between 0 and n_digits (60.000) 
index = np.random.randint(n_digits, size = 50000) 
# set up of samples 
digits_sample = digits[index]
labels_sample = labels[index]

In [0]:
# how much data should be consider? Size = 3000 -> 3000 numbers of the vector // it is faster!
size_test_sample = 3000 
index = np.random.randint(n_test, size = size_test_sample)
test_digits_sample = test_digits[index]
test_labels_sample = test_labels[index]

# Calculating Distances

In [0]:
# setting up the k value 
# k = 2 
#k = 4
 k = 8 
# calculating the distance between each vector in dependency of k and save it in an arra
distances = scipy.spatial.distance.cdist(digits_sample, test_digits_sample, metric='euclid')
# calculates the next k points around the vector
index_nearest = np.argpartition(distances, k, axis = 0)[:k] 


In [66]:
# saves the neighbors of the vector in dependency of k
knn_neighbors = labels_sample[index_nearest] 
# show the neighbors
#print(knn_neighbors)

[[9 8 6 ... 2 1 0]
 [4 8 6 ... 2 1 0]
 [4 8 6 ... 2 1 0]
 ...
 [4 8 6 ... 2 1 0]
 [4 8 6 ... 2 1 0]
 [4 8 6 ... 2 1 0]]


In [67]:
# is looking for the most frequent value in the array/vector
knn_result = stats.mode(knn_neighbors)[0]
knn_result = knn_result[0]
# shows the most frequent value
knn_result

array([4, 8, 6, ..., 2, 1, 0], dtype=uint8)

# Get the global errorrate

In [0]:
errors = 0
# compares the two arrays and is looking for matches of the values? Right = true / Wrong = false
error_rate_matrix = (test_labels_sample == knn_result)
for b in error_rate_matrix:
    if b == False:
        errors += 1



In [69]:
np.unique(test_labels_sample, return_counts=True)

(array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], dtype=uint8),
 array([295, 339, 282, 322, 309, 262, 309, 296, 266, 320]))

In [70]:
# Calculation of the error rate
_rate = (errors / size_test_sample) * 100. 
print("Global error rate is: " +str(_rate) + " %")

Global error rate is: 4.133333333333333 %


*Global error rate* for **k = 2** -> 3.9 %


*Global error rate* for **k = 4** -> 3.5000000000000004%

*Global error rate* for **k = 8** -> 4.133333333333333 %



# Get the individual errorrate

In [0]:
indi_error ={0:0,1:0,2:0,3:0,4:0,5:0,6:0,7:0,8:0,9:0}
total = {0:0,1:0,2:0,3:0,4:0,5:0,6:0,7:0,8:0,9:0}
for i in range(len(test_labels_sample)):
  if test_labels_sample[i]!= knn_result[i]:
    # speicher das die zahl falsch war
    indi_error[test_labels_sample[i]] +=1
    total[test_labels_sample[i]] +=1
  else:
    # speicher das die zahl richtig war
    total[test_labels_sample[i]] +=1

In [0]:
errorrate = {}
for j in range(0,10):
  errorrate[j] = (indi_error[j]/total[j]) * 100
  #print(errorrate[j])


In [75]:
for i in range(len(errorrate)):
  #print("Error rate for number "+ str(i) +" is "+ str(errorrate[i]))

Error rate for number 0 is 1.3559322033898304
Error rate for number 1 is 0.2949852507374631
Error rate for number 2 is 6.73758865248227
Error rate for number 3 is 4.658385093167702
Error rate for number 4 is 3.559870550161812
Error rate for number 5 is 4.198473282442748
Error rate for number 6 is 1.9417475728155338
Error rate for number 7 is 3.7162162162162162
Error rate for number 8 is 9.774436090225564
Error rate for number 9 is 6.25


each error rate for **k = 2**:

* Error rate for number 0 is 1.3559322033898304
* Error rate for number 1 is 0.0
* Error rate for number 2 is 4.609929078014184
* Error rate for number 3 is 4.037267080745342
* Error rate for number 4 is 3.559870550161812
* Error rate for number 5 is 2.6717557251908395
* Error rate for number 6 is 1.2944983818770228
* Error rate for number 7 is 4.391891891891892
* Error rate for number 8 is 10.526315789473683
* Error rate for number 9 is 7.5


each error rate for **k = 4**:


* Error rate for number 0 is 0.3731343283582089
* Error rate for number 1 is 0.5555555555555556
* Error rate for number 2 is 4.398826979472141
* Error rate for number 3 is 4.693140794223827
* Error rate for number 4 is 3.296703296703297
* Error rate for number 5 is 6.177606177606178
* Error rate for number 6 is 2.083333333333333
* Error rate for number 7 is 3.303303303303303
* Error rate for number 8 is 7.000000000000001
* Error rate for number 9 is 3.6544850498338874






each error rate for **k = 8**:
* Error rate for number 0 is 1.3559322033898304
* Error rate for number 1 is 0.2949852507374631
* Error rate for number 2 is 6.73758865248227
* Error rate for number 3 is 4.658385093167702
* Error rate for number 4 is 3.559870550161812
* Error rate for number 5 is 4.198473282442748
* Error rate for number 6 is 1.9417475728155338
* Error rate for number 7 is 3.7162162162162162
* Error rate for number 8 is 9.774436090225564
*Error rate for number 9 is 6.25

# The choice of k

As we can see, the error rate at k = 4 is lowest while k = 8 and k = 2 are worse. 

At k = 4 the best predictions are made.

Choosing the right value of k is a process called parameter tuning, and is important for better accuracy. A lower value of k means that noise will have a bigger impacton the result and a higher k make it more expensive. But a higher value of k has lesser chance of error.

We found out, that this approach could be possible:
 k= sqrt(n) -> where n the number of data points is